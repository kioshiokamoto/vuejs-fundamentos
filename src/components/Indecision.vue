<template>
  <img v-if="img" :src="img" alt="ve" />
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input type="text" placeholder="Realiza una pregunta" v-model="question" />
    <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
    <div>
      <h2 v-if="isValidQuestion">{{ question }}</h2>
      <h1 v-if="isValidQuestion">{{ answer }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  name: "Indecision",
  data() {
    return {
      question: null,
      answer: null,
      img: null,
      isValidQuestion: false,
    };
  },
  methods: {
    async getAnswer() {
      try {
        this.answer = "Pensando...";

        const res = await fetch("https://yesno.wtf/api");
        const { answer, image } = await res.json();

        this.answer =
          answer === "yes" ? "Si" : answer === "no" ? "No!" : "Puede ser";
        this.img = image;
      } catch (error) {
        this.answer="No se pudo cargar del API"
        this.img=null;
      }
    },
  },
  watch: {
    question(value, oldValue) {
      this.isValidQuestion = false;
      console.log({ value });
      if (!value.includes("?")) return;
      this.isValidQuestion = true;
      this.getAnswer();
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>