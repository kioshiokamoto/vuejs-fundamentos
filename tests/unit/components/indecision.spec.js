import { shallowMount, mount } from "@vue/test-utils";
import Indecision from "@/components/Indecision.vue";
describe("Pruebas en componente Indecision", () => {
    let wrapper;
    let clgSpy;

    global.fetch = jest.fn(() =>
        Promise.resolve({
            json: () =>
                Promise.resolve({
                    answer: "yes",
                    forced: false,
                    image: "https://yesno.wtf/assets/yes/2.gif",
                }),
        })
    );

    beforeEach(() => {
        wrapper = shallowMount(Indecision);
        clgSpy = jest.spyOn(console, "log");
        jest.clearAllMocks();
    });
    test("match con snapshot", () => {
        expect(wrapper.html()).toMatchSnapshot();
    });

    test("escribir en input no debe disparar nada ", async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

        const valor = "Hola mundo";
        const input = wrapper.find("input");
        await input.setValue(valor);

        expect(clgSpy).toHaveBeenCalledTimes(1);

        expect(getAnswerSpy).not.toHaveBeenCalled();
    });
    test("escribir simbolo `?` debe disparar el fetch", async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, "getAnswer");

        const valor = "Hola mundo?";
        const input = wrapper.find("input");
        await input.setValue(valor);

        expect(clgSpy).toHaveBeenCalledTimes(1);

        expect(getAnswerSpy).toHaveBeenCalled();
    });
    test("pruebas en getAnswer", async () => {
        await wrapper.vm.getAnswer();
        const img = wrapper.find("img");

        expect(img.exists()).toBeTruthy();
        expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif");
        expect(wrapper.vm.answer).toBe("Si");
    });
    test("pruebas en getAnswer - fallo en API", async () => {
        fetch.mockImplementationOnce(() => Promise.reject("Api failed"));
        await wrapper.vm.getAnswer();


        const img = wrapper.find("img");
        expect(img.exists()).toBeFalsy();
        expect(wrapper.vm.answer).toBe("No se pudo cargar del API");
    });
});
