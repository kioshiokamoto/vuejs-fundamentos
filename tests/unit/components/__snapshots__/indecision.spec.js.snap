// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en componente Indecision match con snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Realiza una pregunta">
  <p>Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <div>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;
